infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentSupervisor.java

```diff
@@ -0,0 +1,181 @@
+package org.corfudb.infrastructure.log;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import static org.corfudb.infrastructure.utils.Persistence.syncDirectory;
+
+/**
+ * The class manages corfu's file segments
+ */
+@Builder
+@Slf4j
+public class SegmentSupervisor {
+    @Getter
+    private final ConcurrentMap<Path, SegmentHandle> channels = new ConcurrentHashMap<>();
+    private final Set<FileChannel> channelsToSync = ConcurrentHashMap.newKeySet();
+
+    @NonNull
+    private final Path logDir;
+
+    /**
+     * Gets the file channel for a particular address, creating it
+     * if is not present in the map.
+     *
+     * @param address The address to open.
+     * @return The FileChannel for that address.
+     */
+    @VisibleForTesting
+    SegmentHandle getSegmentByAddress(long address, AddressSpaceLoader addressSpaceLoader) {
+        long segmentId = address / StreamLogFiles.RECORDS_PER_LOG_FILE;
+
+        Path filePath = logDir.resolve(segmentId + ".log");
+
+        SegmentHandle handle = channels.computeIfAbsent(filePath, missingFilePath -> {
+            FileChannel writeCh = null;
+            FileChannel readCh = null;
+
+            try {
+                writeCh = getChannel(missingFilePath, false);
+                readCh = getChannel(missingFilePath, true);
+
+                SegmentHandle segment = new SegmentHandle(segmentId, writeCh, readCh, missingFilePath);
+                // The first time we open a file we should read to the end, to load the
+                // map of entries we already have.
+                // Once the segment address space is loaded, it should be ready to accept writes.
+                addressSpaceLoader.load(segment);
+                return segment;
+            } catch (IOException e) {
+                log.error("Error opening file {}", missingFilePath, e);
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw new IllegalStateException(e);
+            } catch (RuntimeException ex) {
+                //Prevents file resources leaks in case of any RuntimeException.
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw ex;
```

Comment:
If this is not expected, then its a good idea to log an error.

Reply:
If we throw an exception then no need for logs. In the other case, we would have a lot of logs on different levels but the logs will be useless. Since the exception will be handled on upper levels, we expect that any necessary logging will happen on those levels